Support TypeScript 4.1

clydin opened the pull request:
This change updates the framework to support TypeScript 4.1.

clydin committed:
feat(compiler-cli): add support for using TypeScript 4.1

This change enables projects to be built with TypeScript 4.1.  Support for TypeScript 4.0 is also retained.

clydin committed:
build: support building with TypeScript 4.1

TypeScript 4.1 is now used to build and test within the repository.

clydin committed:
fixup! build: support building with TypeScript 4.1

JoostK committed:
fixup! build: support building with TypeScript 4.1

JoostK committed:
fixup! build: support building with TypeScript 4.1

JoostK committed:
fixup! build: support building with TypeScript 4.1

JoostK commented:
@googlebot I consent.

JoostK commented on packages/compiler-cli/ngcc/src/host/commonjs_umd_utils.ts:
```diff
@@ -78,7 +78,7 @@ export function findNamespaceOfIdentifier(id: ts.Identifier): ts.Identifier|null
     null {
   const symbol = checker.getSymbolAtLocation(id) || null;
-  const declaration = symbol && symbol.valueDeclaration;
+  const declaration = symbol?.valueDeclaration ?? symbol?.declarations?.[0];
```
This change seems to have been due to microsoft/TypeScript#39770, where the variable declaration is now treated as an alias declaration which no longer has a value declaration.

alan-agius4 reviewed (commented):
Can we add an integration tests like https://github.com/angular/angular/tree/master/integration/typings_test_ts40

clydin committed:
fixup! build: support building with TypeScript 4.1

clydin committed:
test(compiler-cli): add TypeScript 4.1 typings integration test

This change adds a typings test identical to the TypeScript 4.0 test but using TypeScript 4.1 instead.

gkalpak reviewed (approved):
LGTM for `dev-infra`, `docs-infra`, `fw-upgrade`.

From a quick look, I would expect the [`paths` without `baseUrl`][1] breaking change to affect `ngcc`, because we iirc we assume `baseUrl` will be present. For example, `ngcc` [seems to ignore `paths`][2] if there is no `baseUrl`.

Reviewed-for: dev-infra, docs-infra, fw-upgrade

[1]: https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#paths-without-baseurl
[2]: https://github.com/angular/angular/blob/3e1e5a15ba1d5a08daf9179083e3394df10299bb/packages/compiler-cli/ngcc/src/path_mappings.ts#L22

JoostK commented on packages/compiler-cli/test/transformers/program_spec.ts:
```diff
@@ -297,15 +297,27 @@ describe('ng program', () => {
+            setTempProgramHandlerForTest(program => {
+              structureReuse = tsStructureIsReused(program);
+            });
+            compile();
```
That would work, but we'd have to be sure that it is exception safe so the `afterEach` is still nice to have IMO (or wrap it in try-finally)

clydin committed:
fixup! build: support building with TypeScript 4.1

clydin commented on aio/package.json:
```diff
@@ -164,7 +164,7 @@
     "ts-node": "^8.4.1",
     "tslint": "~6.1.0",
-    "typescript": "~4.0.2",
+    "typescript": "~4.1.2",
```
good point.  Reverted.

clydin commented on integration/typings_test_ts41/include-all.ts:
```diff
@@ -0,0 +1,67 @@
+import * as core from '@angular/core';
+import * as coreTesting from '@angular/core/testing';
+import * as elements from '@angular/elements';
+import * as forms from '@angular/forms';
```
good question.  Wasn't there in the 4.0 tests but we can add it.

gkalpak commented on integration/typings_test_ts41/include-all.ts:
```diff
@@ -0,0 +1,69 @@
+import * as commonHttpTesting from '@angular/common/testing';
+import * as compiler from '@angular/compiler';
+import * as compilerTesting from '@angular/compiler/testing';
+import * as core from '@angular/core';
```
What about `@angular/core/global` 😇:

```suggestion
import * as core from '@angular/core';
import * as coreGlobal from '@angular/core/global';
```

(And also export below.)

clydin replied:
Thinking about this some more, I think for the purposes of this PR (enabling TS 4.1 use) we should keep the 4.0 and 4.1 tests consistent.
And then a follow-up PR can fix both (potentially in a more long-term maintainable fashion).  Components, for instance, has a more generic approach to the concept: https://github.com/angular/components/tree/bb2bb9c3065ec1a7a845de072305893a7378383a/integration/ts-compat

gkalpak commented on integration/typings_test_ts41/include-all.ts:
```diff
@@ -0,0 +1,69 @@
+import * as coreTesting from '@angular/core/testing';
+import * as elements from '@angular/elements';
+import * as forms from '@angular/forms';
+import * as localize from '@angular/localize';
```
Don't we also need to import from `@angular.localize/init`?

```suggestion
import * as localizeInit from '@angular/localize/init';
```

(And also export below.)

jpike88 commented:
Just checking... I got the error `Internal Error: The structure of the program changed during codegen` when attempting to use typescript 4.1.2 with angular cli 11.0.2. Is this expected, and covered in the PR's fixes? I can't seem to find anyone mentioning that kind of error, and not sure what to make of it.

JoostK commented:
> Just checking... I got the error `Internal Error: The structure of the program changed during codegen` when attempting to use typescript 4.1.2 with angular cli 11.0.2. Is this expected, and covered in the PR's fixes? I can't seem to find anyone mentioning that kind of error, and not sure what to make of it.

Yes that is expected for ViewEngine compilations, but not Ivy. And yes that is covered in this PR.

JiaLiPassion reviewed (approved):
LGTM
Reviewed-for: zone-js

IgorMinar reviewed (approved):
LGTM! Thank you!

Reviewed-for: global-approvers
